sos/policies/auth/deviceauth.py

```diff
@@ -0,0 +1,313 @@
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+import logging
+try:
+    import requests
+    REQUESTS_LOADED = True
+except ImportError:
+    REQUESTS_LOADED = False
+import time
+from datetime import datetime, timedelta
+import os
+from sos.policies.auth.keyring import Key, KeyNotFoundError, KeyRevokedError
+from sos.utilities import is_executable
+
+DEVICE_AUTH_CLIENT_ID = "sos-tools"
+GRANT_TYPE_DEVICE_CODE = "urn:ietf:params:oauth:grant-type:device_code"
+DATETIME_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+
+logger = logging.getLogger("sos")
+
+
+def try_read_refresh_token():
+    """Try to read locally stored refresh token
+
+    Returns:
+        str: Returns ODIC refresh token if found otherwise None
+    """
+    try:
+        AuthKey = Key()
+        key_refresh_token = AuthKey.search('sos-tools_refresh_token')
+        key_username = AuthKey.search('sos-tools-user')
+    except KeyNotFoundError:
+        logger.info("Refresh token does not exist in keyring or is expired.")
+        return None
+    except KeyRevokedError:
+        logger.error("Key was revoked. You will need to upload the "
+                     "sos report manually.")
+        raise
+    except Exception as e:
+        logger.error("Error encountered while accessing keyring. {}".format(e))
+        return None
+    if key_username.data.decode() != os.getlogin():
+        return None
+
+    return key_refresh_token.data.decode()
+
+
+class DeviceAuthorizationClass:
+    """
+    Device Authorization Class
+    """
+
+    def __init__(self, client_identifier_url, token_endpoint):
+
+        self.client_identifier_url = None
+        self.token_endpoint = None
+        self._access_token = None
+        self._access_expires_at = None
+        self._refresh_token = None
+        self._refresh_expires_at = None
+        self._refresh_expires_in = None
+        self._user_verification_url = None
+        self.__device_code = None
+        self.AuthKey = Key()
+
+        # Lets check first if we have keyctl installed so we can
+        # store the token in the keyring
+        if not is_executable('keyctl'):
+            raise Exception("keyctl tool is not installed"
+                            " and is required to store auth tokens.")
+
+        self.client_identifier_url = client_identifier_url
+        self.token_endpoint = token_endpoint
+        self._use_device_code_grant()
+
+    def _use_device_code_grant(self):
+        """
+        Start the device auth flow. First check for the refresh token stored in
+        the session keyring. If they are not stored
+        or are invalid, request new device code. If the stored refresh token is
+        valid, use it to get new access_token
+
+        """
+        try:
+            stored_refresh_token = try_read_refresh_token()
+        except KeyRevokedError as e:
```

Comment:
Why is there same "expect" as inside of try_read_refresh_token method? It is already checked inside the method so if there is try/except i would at least expect more generic exception.

Reply:
You are absolutely right. I'll fix this in the next push